### ISSUE TITLE
Generating batch outputs?

### ISSUE THREAD

**Issue by ron-vnai** [Lightning-AI/litgpt#345]
Hi, I wonder if it is possible to generate responses over a batch of input queries?

The scenario I consider is having a batch of different input queries (say all have the same number of tokens), and generating output responses for the whole batch in an efficient manner. 

Naturally, I can modify the "generate/base.py" generate function to support batch inputs naively. However, I look for efficient implementation. 

Thanks!

**Comment by carmocca** [comment]
Sorry, this is not implemented at the moment for simplicity in understanding the generation code, (it's inherited from nanoGPT)

**Comment by yacineMTB** [comment]
I think it's supported (by model.py). Try left padding the input sequences across each sequence in the batch. e.g. idx [B, T], where T is the longest length across all sequences, and the smaller ones get padded up.

You should be able to do something like

```
padded = torch.ones(T, dtype=dtype, device=device) * tokenizer.bos_id
padded[-len(sequence):] = sequence.clone().detach()
```

(i think)

**Comment by MF-FOOM** [comment]
why left pad — wouldn't you want the padding tokens to be at the end so the generations aren't all conditioned on the padding tokens? ig you can add an attention mask to avoid but seems easier just to right pad and sample logits from the last token before padding?

**Comment by MF-FOOM** [comment]
> Try left padding the input sequences across each sequence in the batch. 

just remember to implement an [attention mask](https://huggingface.co/docs/transformers/main/glossary#attention-mask) so the model ignores the padding tokens or your outputs will be conditioned on the padding tokens which may give you strange outputs

**Comment by MF-FOOM** [comment]
also wait wouldn't left padding screw up the positional encoding, even if you attention mask?

**Comment by yacineMTB** [comment]
Ah, cheers!

Attention masking should be straightforward, two things need to be done:
- change build_mask_cache to mask based on the padding input: `(B, T, block_size, block_size)`. Should carry through
- Use a large negative number instead of a boolean mask on `scaled_dot_product_attention`, to work around a bug on pytorch's implementation. (see https://github.com/pytorch/pytorch/issues/103749)

I think you're right about the left padding needing to be accounted for in the positional encoding. After adding the mask, things got thrown off - which makes sense. I think that [this is happening here](https://github.com/facebookresearch/llama/blob/main/llama/model.py#L469-L470)? And, also, here, [on huggingface's llama impl](https://github.com/huggingface/transformers/blob/0afa5071bd84e44301750fdc594e33db102cf374/src/transformers/models/idefics/modeling_idefics.py#L537-L544)

**Comment by yacineMTB** [comment]
not shilling, but this does batching :)

https://github.com/yacineMTB/just-large-models/blob/master/llama.py

**Comment by carmocca** [comment]
It looks like the Llama 2 batched inference code does right-padding? https://github.com/facebookresearch/llama/blob/main/llama/generation.py#L162-L176

I'm not 100% familiar with the advantages of left vs right so if one of you has a good resource on this, I'd appreciate it if you could share it

**Comment by carmocca** [comment]
From what I understand, right padding will not require creating an attention mask (so you can keep using flash attention), but then one cannot simply `-1` here: https://github.com/Lightning-AI/lit-gpt/blob/main/generate/base.py#L62

**Comment by rasbt** [comment]
Another advantage of right-padding is that it is (more) compatible with models that have an absolute positional embedding

**Comment by jinulee-v** [comment]
Hello there, I have implemented a **very naive** version of batched inference compatible to `lit_gpt`. The most critical problem here is that it does not support `kv_cache` . I inspected the KV cache code in `lit_gpt/base.py`, but had no idea how to deal with variable length prompts neatly since `cos` and `sin` are 1D-tensors. Maybe expanding this into 2D could be a solution...

Paste this gist into `generate/base.pt` :
https://gist.github.com/jinulee-v/d7dcfe9ad1280acc69c279fcbfcdfd22

- Can : generate from variable-length prompts with right-padding 
- Can't : use kv_cache.
